### PR TITLE
Free the char 'code' after g_malloc0 in gnc-flicker-gui.c

### PR DESCRIPTION
**Pull request by Bob-IT**
I think this is correct but do not like changing stuff I can not test.
@jralls Do these changes look correct ?

Change flicker_data to return a guint so that the char code can be
freed in the function which also removes the need to use g_malloc0
twice.

### CODE REVIEW

**Comment by CWehli** [comment]
@Bob-IT, @jralls: I checked the correction and the TAN generation continues to work.

**Comment by jralls** [comment]
@Bob-IT Good catch. Let's go one further and replace
```

     /* prepend synchronization identifier */
     const char pre[] = {'0', 'F', 'F', 'F'};
     guint challenge_length;
     size_t len = sizeof (pre) + strlen (challenge) + 1;
     char* code = (char*)g_malloc0 (len);
     memcpy (code, pre, sizeof (pre));
     memcpy (code + sizeof (pre), challenge, strlen (challenge));
```
with
```
     /* prepend synchronization identifier */
    char* code = g_strdup_printf ("0FFF%s", challenge);
    guint challenge_length = strlen (code);
```

**Comment by CWehli** [comment]
@Bob-IT Still works if you also replace 'len' with 'challenge_length' in the for loop.

**Comment by Bob-IT** [comment]
@CWehli I have made the suggested changes can you make sure they are still OK before I push tomorrow.
I forgot to change the commit text so will do that before I push.

File: gnucash/import-export/aqb/gnc-flicker-gui.c [hunk 1]
```diff
@@ -114,38 +114,37 @@ get_num (const char ch)
...
-     for (guint i = 0; i < len - 1; i += 2)
-     {
-         guint val1 = get_num (code[i]);
-         guint val2 = get_num (code[i+1]);
-
-         memcpy (&bitarray[i], bits[val2], sizeof(bits[val2]));
-         memcpy (&bitarray[i+1], bits[val1], sizeof(bits[val1]));
-     }
-     return code;
+    for (guint i = 0; i < challenge_length - 1; i += 2)
```
Comment (jralls) [inline gnucash/import-export/aqb/gnc-flicker-gui.c:137]: I think this will leave off the last character as `strlen` doesn't count the terminal `'\0'`.

Comment (CWehli) [reply gnucash/import-export/aqb/gnc-flicker-gui.c:137]: @Bob-IT, @jralls: I have made bank transfers and the TAN is accepted.

I also changed the counter from "challenge_length-1" to "challenge_length-0" and the generation of the TAN still works.

Comment (Bob-IT) [reply gnucash/import-export/aqb/gnc-flicker-gui.c:137]: @CWehli So just need to change it to challenge_length, if that is correct I will do that tomorrow before I push.

Comment (jralls) [reply gnucash/import-export/aqb/gnc-flicker-gui.c:137]: > I also changed the counter from "challenge_length-1" to "challenge_length-0" and the generation of the TAN still works.

It's interesting that it worked without the last character. It makes one wonder if there's some padding included in the challenge.

**Comment by CWehli** [comment]
@jralls , @Bob-IT : So, I have tested it again.
The last two numbers are a check digit. If I shorten "challenge_length" by 0 or 1 in the FOR loop of the counter, the TAN generation works correctly. If I subtract 2 then it no longer works. 

I suggest running the FOR loop up to the entire length of strlen(), i.e. using "for (i = 0, i < challenge_length, i += 2)".

**Comment by jralls** [comment]
@CWehli Makes sense, thanks.

**Comment by Bob-IT** [comment]
@CWehli Thank you for testing. In an unrelated matter, do you suffer from bug https://bugs.gnucash.org/show_bug.cgi?id=768547
I added a possible fix but can not test, no worries if not affected.

**Comment by CWehli** [comment]
@Bob-IT : mhh, I don't remember any such behaviour. At least your improvement does not seem to have any negative effects.

**Pull request merged**